### PR TITLE
Aggregate tests

### PR DESCRIPTION
**Pull request by NeilMenne**
tests will be simpler once the last major pieces are put in place, but for now, I am manually starting and stopping for the SampleAggregate.

Happy reading!

### CODE REVIEW

File: lib/event_store/aggregate.ex [hunk 1]
```diff
@@ -43,59 +56,95 @@ defmodule EventStore.Aggregate do
...
+        end
+      end
 
-      def use_snapshot(_, snapshot), do: snapshot.body
+      defp command_update(%{sequence: a} = agg, %{sequence: c}) when c > a,
+        do: update_aggregate(agg)
+      defp command_update(agg, _), do: agg
+
+      def handle_events(agg, command, events) do
+        case Store.commit_events!(events) do
```
Comment (keathley) [inline lib/event_store/aggregate.ex]: If its a `!` function then it shouldn't be returning a match it should just raise an exception.

File: lib/event_store/store/adapter.ex [hunk 2]
```diff
@@ -1,17 +1,54 @@
...
+  stricter evaluation rules for commands. If the events could not be committed
+  for any other reason, the storage mechanism should raise an appropriate
+  exception.
+  """
+  @callback commit_events!([Event.t]) :: :ok
                                        | {:error, :retry_command}
                                        | :no_return
 
+  @doc """
+  Snapshots are committed iff the proposed version is newer than the version
```
Comment (keathley) [inline lib/event_store/store/adapter.ex]: Small typo here `iff`.

File: lib/event_store/store/postgres.ex [hunk 3]
```diff
@@ -31,25 +38,36 @@ defmodule EventStore.Store.Postgres do
...
 
-  def get_events(aggregate_id, seq \\ 0) do
-    Repo.all(
-      from e in Event,
-      where: e.sequence > ^seq,
-      where: e.aggregate_id == ^aggregate_id,
-      select: e
-    )
+  def get_events(aggregate_id, seq, options) do
+    base = from e in Event
```
Comment (keathley) [inline lib/event_store/store/postgres.ex]: I might make this a function like `def event_query, do: from e in Event`.

File: lib/event_store/store/postgres.ex [hunk 4]
```diff
@@ -31,25 +38,36 @@ defmodule EventStore.Store.Postgres do
...
+  end
+
+  def for_aggregate(query, agg_id) do
+    from r in query,
+      where: r.aggregate_id == ^agg_id,
+      select: r
+  end
+
+  def get_snapshot(aggregate_id, seq, options) do
+    base = from s in Snapshot
```
Comment (keathley) [inline lib/event_store/store/postgres.ex]: Same as the above comment.

File: lib/event_store/store/postgres.ex [hunk 5]
```diff
@@ -58,16 +76,16 @@ defmodule EventStore.Store.Postgres do
...
-         {:ok, %{} = res} -> Map.values(res)
+         {:ok, _} -> :ok
        end
   end
 
   defp append_changeset(cs, mult),
-    do: Multi.insert(mult, changeset_key(cs), cs, returning: true)
+    do: Multi.insert(mult, changeset_key(cs), cs)
 
   defp retry_error({:error, _, %{errors: [sequence: {:dupe_seq_agg, _}]}, _}),
```
Comment (keathley) [inline lib/event_store/store/postgres.ex]: Why not just do this pattern match in the case on line 78 and pass in the list of errors?

File: lib/event_store/store/postgres.ex [hunk 6]
```diff
@@ -58,16 +76,16 @@ defmodule EventStore.Store.Postgres do
...
     do: {:error, :retry_command}
-  defp retry_error(err), do: raise EventStore.StoreError.exception(err)
```
Comment (keathley) [inline lib/event_store/store/postgres.ex]: I'm not sure how I feel about raising here. Is it possible to return `{:error, StoreError.exception(err)}` and raise it at the call site?

File: mix.exs [hunk 7]
```diff
@@ -33,18 +32,15 @@ defmodule EventStore.Mixfile do
...
     [
-      {:phoenix, "~> 1.3.0"},
-      {:phoenix_pubsub, "~> 1.0"},
-      {:phoenix_ecto, "~> 3.2"},
+      {:ecto, "~> 2.2"},
       {:postgrex, ">= 0.0.0"},
-      {:gettext, "~> 0.11"},
-      {:cowboy, "~> 1.0"},
+      {:poison, "~> 3.0"},
+      {:ecto_hlclock, git: "git@github.com:tonic-sys/ecto_hlclock.git"},
```
Comment (keathley) [inline mix.exs]: We should go ahead and push a v0.1 of this to hex

File: test/event_store/aggregate_test.exs [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  alias EventStore.Command
+  alias EventStore.SampleAggregate
+
+  setup do
+    Application.put_env(
+      :event_store,
+      :storage_adapter,
+      EventStore.Store.InMemory
+    )
+    EventStore.Store.InMemory.reset
```
Comment (keathley) [inline test/event_store/aggregate_test.exs]: Does this need to be reset between each test or before all of these tests?

File: lib/event_store/aggregate.ex [hunk 9]
```diff
@@ -17,13 +17,26 @@ defmodule EventStore.Aggregate do
...
-  @callback apply_event(state :: any, event :: Event.t) :: any
-  @callback use_snapshot(curr :: any, snapshot :: Snapshot.t) :: any
+  defstruct [:id, :sequence, :state]
+  @type t :: %__MODULE__{
+    id: HLClock.Timestamp.t,
+    sequence: integer,
+    state: any
+  }
+
+  @type agg :: __MODULE__.t
```
[the remark this answers is not in the document]
Comment (NeilMenne) [reply lib/event_store/aggregate.ex]: `Aggregate`

[the remark this answers is not in the document]
Comment (keathley) [reply lib/event_store/aggregate.ex]: You can just use `t` then.

File: test/event_store/aggregate_test.exs (hunk 8 above)
Comment (NeilMenne) [reply test/event_store/aggregate_test.exs]: between each test due to the occasional duplicate timestamp being generated as the aggregate ID.

File: mix.exs (hunk 7 above)
Comment (NeilMenne) [reply mix.exs]: good call. will do

**Pull request merged**